File: shared/types/kb.ts
/**
 * 知识库域（F-KB）wire 类型：与 backend/app/schemas/kb.py CamelModel 单一真相源对齐。
 */

/** 知识库设置响应（管理后台读写共用）。 */
export interface ApiKbSettingsResponse {
  hotwords: string[]
  segmentMaxSeconds: number
  asrConcurrency: number
  topK: number
  autoApprovePoints: boolean
  unitPrices: Record<string, number>
  embeddingConfigId: number | null
  cleanModelId: number | null
  extractModelId: number | null
  visionModelId: number | null
  authorizedUserIds: number[]
  updatedAt: string | null
}

/** 知识库设置保存请求（全部可选，仅提交的字段更新）。 */
export interface ApiKbSettingsUpdateRequest {
  hotwords?: string[]
  segmentMaxSeconds?: number
  asrConcurrency?: number
  topK?: number
  autoApprovePoints?: boolean
  unitPrices?: Record<string, number>
  embeddingConfigId?: number | null
  cleanModelId?: number | null
  extractModelId?: number | null
  visionModelId?: number | null
  authorizedUserIds?: number[]
}

/** 素材类型（书的 episodeNo 恒为 null）。 */
export type ApiKbMediaKind = 'video' | 'audio' | 'book'

/** 素材处理状态机：uploaded → awaiting_cost → queued → processing → done/failed。 */
export type ApiKbProcessStatus =
  | 'uploaded'
  | 'awaiting_cost'
  | 'queued'
  | 'processing'
  | 'done'
  | 'failed'

/** 知识源视图。 */
export interface ApiKbSourceResponse {
  id: number
  sourceType: 'course' | 'book'
  name: string
  author: string | null
  description: string | null
  enabled: boolean
  storageBytes: number
  pendingCleanupBytes: number
  deletedAt: string | null
  createdAt: string
  updatedAt: string
}

/** 新建知识源请求。 */
export interface ApiKbSourceCreateRequest {
  sourceType: 'course' | 'book'
  name: string
  author?: string | null
  description?: string | null
  enabled?: boolean
}

/** 编辑知识源请求（全部可选，仅提交的字段更新）。 */
export interface ApiKbSourceUpdateRequest {
  name?: string
  author?: string | null
  description?: string | null
  enabled?: boolean
}

/** 素材视图。 */
export interface ApiKbMediaResponse {
  id: number
  sourceId: number
  mediaKind: 'video' | 'audio' | 'book'
  episodeNo: number | null
  title: string | null
  fileName: string
  relativePath: string | null
  fileSize: number | null
  fileHash: string | null
  durationSeconds: number | null
  pageCount: number | null
  processStatus: ApiKbProcessStatus
  processError: string | null
  processMeta: Record<string, unknown>
  editedAt: string | null
  deletedAt: string | null
  pointCount: number
  dirtyCount: number
  createdAt: string
  updatedAt: string
}

/** 批量登记的单个文件（目录结构由 relativePath 保留）。 */
export interface ApiKbMediaInitItem {
  fileName: string
  relativePath: string
  size: number
  hash: string
  mediaKind: 'video' | 'audio' | 'book'
  episodeNo?: number | null
  title?: string | null
  durationSeconds?: number | null
  pageCount?: number | null
}

/** 批量登记请求。 */
export interface ApiKbMediaInitRequest {
  items: ApiKbMediaInitItem[]
}

/** 批量登记结果（每文件一行，与请求 items 等长同序；conflictWith 非空表示该文件与库内已有素材重复、被跳过）。 */
export interface ApiKbMediaInitResult {
  mediaId: number | null
  fileName: string
  cosKey: string | null
  uploadUrl: string | null
  conflictWith: string | null
}

export interface ApiKbMediaInitResponse {
  items: ApiKbMediaInitResult[]
}

/** 分片上传会话创建/续传请求（partSize/partCount 由前端按文件大小计算）。 */
export interface ApiKbUploadSessionRequest {
  partSize: number
  partCount: number
  resumeUploadId?: string | null
}

/** 已完成分片（服务端 list_parts 真相）。 */
export interface ApiKbUploadSessionPart {
  partNumber: number
  etag: string
  size: number
}

/** 待上传分片的预签名 PUT URL。 */
export interface ApiKbUploadSessionPartUrl {
  partNumber: number
  url: string
}

/** 分片上传会话视图：completedParts 已传分片，partUrls 仅覆盖缺失分片。 */
export interface ApiKbUploadSessionResponse {
  mediaId: number
  uploadId: string
  partSize: number
  partCount: number
  completedParts: ApiKbUploadSessionPart[]
  partUrls: ApiKbUploadSessionPartUrl[]
}

/** 素材信息修正请求（全部可选）。 */
export interface ApiKbMediaPatchRequest {
  episodeNo?: number | null
  title?: string
  durationSeconds?: number | null
  pageCount?: number | null
}

/** 费用预估请求（素材必须属于该知识库）。 */
export interface ApiKbCostEstimateRequest {
  sourceId: number
  mediaIds: number[]
}

/** 单素材费用分项（书在批次 C 接入前为 0 项）。 */
export interface ApiKbCostEstimateItem {
  mediaId: number
  title: string
  mediaKind: string
  durationSeconds: number | null
  asrCost: number
  cleanTokens: number
  estimatedCost: number
}

/** 费用预估视图（预估即把 uploaded 素材推进到 awaiting_cost）。 */
export interface ApiKbCostEstimateResponse {
  currency: string
  items: ApiKbCostEstimateItem[]
  total: number
}

/** 费用确认请求（awaiting_cost → queued）。 */
export interface ApiKbConfirmCostRequest {
  mediaIds: number[]
}

/** 费用确认结果（本次实际入队的素材）。 */
export interface ApiKbConfirmCostResponse {
  queuedIds: number[]
}

/** 文稿分段视图（时间轴只读，编辑只覆盖 text）。 */
export interface ApiKbTranscriptSegmentView {
  seqNo: number
  text: string
  startMs: number | null
  endMs: number | null
  pageStart: number | null
  pageEnd: number | null
}

/** 单集文稿读取。 */
export interface ApiKbTranscriptResponse {
  mediaId: number
  editedAt: string | null
  segments: ApiKbTranscriptSegmentView[]
}

/** 文稿批量保存请求（按 seqNo 覆盖文本）。 */
export interface ApiKbTranscriptUpdateRequest {
  segments: { seqNo: number; text: string }[]
}

/** 文稿保存结果（updatedCount>0 即触发脏传播）。 */
export interface ApiKbTranscriptSaveResponse {
  mediaId: number
  editedAt: string | null
  updatedCount: number
}

// ---------------------------------------------------------------------------
// 知识审核（F-KB-03）
// ---------------------------------------------------------------------------

/** 知识点类型。 */
export type ApiKbPointType = 'concept' | 'theorem' | 'method' | 'discipline' | 'case'

/** 审核状态（仅 published 参与检索）。 */
export type ApiKbPointStatus = 'draft' | 'published' | 'rejected'

/** 章节树节点（id 服务端按位置生成："1"/"1.2"）。 */
export interface ApiKbChapterNode {
  id: string
  title: string
  children: ApiKbChapterNode[]
}

/** 目录树读取（draft 供编辑，published 为生效版本）。 */
export interface ApiKbChaptersResponse {
  draft: ApiKbChapterNode[] | null
  published: ApiKbChapterNode[] | null
}

/** 目录树整棵发布请求。 */
export interface ApiKbChaptersPublishRequest {
  chapters: ApiKbChapterNode[]
}

/** 知识点卡片视图（episodeNo/mediaTitle 为 join 冗余，供原文脚注展示）。 */
export interface ApiKbKnowledgePoint {
  id: number
  sourceId: number
  mediaId: number
  episodeNo: number | null
  mediaTitle: string | null
  pointType: ApiKbPointType
  title: string
  body: string
  termDefinition: string | null
  applicableScene: string | null
  excerpt: string
  startMs: number | null
  endMs: number | null
  pageStart: number | null
  pageEnd: number | null
  relatedIds: number[]
  chapterPath: string[]
  status: ApiKbPointStatus
  needsReview: boolean
  reviewNote: string | null
  reviewedAt: string | null
  createdAt: string
  updatedAt: string
}

/** 审核工作台状态计数。 */
export interface ApiKbPointCounts {
  draft: number
  published: number
  rejected: number
  needsReview: number
}

/** 知识点分页列表。 */
export interface ApiKbPointListResponse {
  items: ApiKbKnowledgePoint[]
  total: number
  counts: ApiKbPointCounts
}

/** 白名单修订请求（excerpt/时间码定位字段服务端拒绝为 422）。 */
export interface ApiKbPointPatchRequest {
  title?: string
  pointType?: ApiKbPointType
  body?: string
  termDefinition?: string | null
  applicableScene?: string | null
  chapterPath?: string[]
}

/** 图片视觉描述状态。 */
export type ApiKbDescribeStatus = 'pending' | 'processing' | 'done' | 'failed'

/** 图片资产视图（书嵌图 + 课程关键帧；thumbUrl 为短时效签名）。 */
export interface ApiKbImageAsset {
  id: number
  sourceId: number
  mediaId: number
  pageNo: number | null
  startMs: number | null
  endMs: number | null
  thumbUrl: string | null
  describeStatus: ApiKbDescribeStatus
  describeAttempts: number
  textInImage: string | null
  caption: string | null
  visionDescription: string | null
  indexExcluded: boolean
  createdAt: string
}

/** 图片资产索引排除开关。 */
export interface ApiKbImageExcludedRequest {
  indexExcluded: boolean
}

/** 图片资产分页列表。 */
export interface ApiKbImageListResponse {
  items: ApiKbImageAsset[]
  total: number
}

/** 人工新增请求（status=draft 走同一审核流）。 */
export interface ApiKbPointCreateRequest {
  mediaId: number
  pointType: ApiKbPointType
  title: string
  body: string
  termDefinition?: string
  applicableScene?: string
  excerpt?: string
  startMs?: number
  endMs?: number
  pageStart?: number
  pageEnd?: number
  chapterPath?: string[]
}

/** 驳回请求（理由入 review_note）。 */
export interface ApiKbPointRejectRequest {
  reason?: string | null
}

/** 重复草稿合并请求（source 行硬删，related 并集进目标）。 */
export interface ApiKbPointsMergeRequest {
  targetId: number
  sourceIds: number[]
}

/** 批量通过请求（已发布/不存在幂等跳过）。 */
export interface ApiKbPointsBatchApproveRequest {
  ids: number[]
}

/** 批量通过结果统计（skipped = 已发布或不存在的卡）。 */
export interface ApiKbBatchApproveResult {
  approved: number
  skipped: number
}

// ---------------------------------------------------------------------------
// 混合检索（消费侧 /kb/search）
// ---------------------------------------------------------------------------

/** 案例卡片关联帧（命中时间窗内就近取帧，thumbUrl 短时效签名）。 */
export interface ApiKbSearchFrameHit {
  id: number
  startMs: number | null
  thumbUrl: string | null
  caption: string | null
}

/** 知识卡片命中（PG 水合全字段；score 为 RRF 融合分）。 */
export interface ApiKbSearchPointHit {
  id: number
  sourceId: number
  mediaId: number
  mediaKind: ApiKbMediaKind
  episodeNo: number | null
  mediaTitle: string | null
  pointType: ApiKbPointType
  title: string
  body: string
  termDefinition: string | null
  applicableScene: string | null
  excerpt: string
  chapterPath: string[]
  relatedIds: number[]
  startMs: number | null
  endMs: number | null
  pageStart: number | null
  pageEnd: number | null
  score: number
  frames: ApiKbSearchFrameHit[]
}

/** 原文分段命中（seekMs = max(0, startMs − 4s) 前滚起播点）。 */
export interface ApiKbSearchSegmentHit {
  id: number
  sourceId: number
  mediaId: number
  mediaKind: ApiKbMediaKind
  episodeNo: number | null
  mediaTitle: string | null
  text: string
  startMs: number | null
  endMs: number | null
  seekMs: number | null
  score: number
}

/** 图片命中（课程帧带时间码、书嵌图带页码）。 */
export interface ApiKbSearchImageHit {
  id: number
  sourceId: number
  mediaId: number
  mediaKind: ApiKbMediaKind
  episodeNo: number | null
  pageNo: number | null
  startMs: number | null
  textInImage: string | null
  caption: string | null
  thumbUrl: string | null
  score: number
}

/** 混合检索结果（三类命中独立列表；degraded 标记降级原因）。 */
export interface ApiKbSearchResponse {
  query: string
  degraded: string | null
  points: ApiKbSearchPointHit[]
  segments: ApiKbSearchSegmentHit[]
  images: ApiKbSearchImageHit[]
}

/** 发布态章节树（消费侧导航，不含 draft）。 */
export interface ApiKbPublishedChaptersResponse {
  chapters: ApiKbChapterNode[]
}

/** 章节浏览卡片（确定性清单非相关性命中，无 score/frames）。 */
export interface ApiKbBrowsePointItem {
  id: number
  sourceId: number
  mediaId: number
  mediaKind: ApiKbMediaKind
  episodeNo: number | null
  mediaTitle: string | null
  pointType: ApiKbPointType
  title: string
  body: string
  termDefinition: string | null
  applicableScene: string | null
  excerpt: string
  chapterPath: string[]
  relatedIds: number[]
  startMs: number | null
  endMs: number | null
  pageStart: number | null
  pageEnd: number | null
}

/** 章节卡片清单（浏览路径：PG 真相源 + episode/startMs 确定性排序分页）。 */
export interface ApiKbChapterPointsResponse {
  total: number
  page: number
  pageSize: number
  points: ApiKbBrowsePointItem[]
}

/** 播放凭证（prev/nextMediaId 供播放器切集；书素材携带 pageCount）。 */
export interface ApiKbPlaybackToken {
  token: string
  expiresIn: number
  mediaId: number
  /** 同时效预签名 GET 直链（video/audio）；book 为 None（书页走 token 代理渲染水印）。 */
  streamUrl: string | null
  prevMediaId: number | null
  nextMediaId: number | null
  pageCount: number | null
}

/** 消费侧知识库条目（启用中库的最小投影）。 */
export interface ApiKbConsumerSource {
  id: number
  name: string
  sourceType: 'course' | 'book'
}

/** 图片原图短时效预签名 URL（≤15min，点击原图时签发）。 */
export interface ApiKbImageUrl {
  url: string
  expiresIn: number
}

/** 用量聚合：kb_* token 分项（feature ∈ kb_clean/kb_extract/kb_vision/kb_embed）。 */
export interface ApiKbUsageTokenItem {
  feature: string
  modelName: string | null
  calls: number
  promptTokens: number
  completionTokens: number
  totalTokens: number
  /** 仅视觉分项按 vlmPerImage × 调用次数估算 */
  estimatedCost: number | null
}

/** 用量聚合：ASR 时长与费用（按时长计费，不走 token 台账）。 */
export interface ApiKbUsageAsr {
  mediaCount: number
  audioSeconds: number
  /** duration 登记口径合计（预估对照） */
  estimatedSeconds: number
  costPerHour: number | null
  cost: number | null
}

/** 建库用量聚合响应（token 分项 + ASR + 预估对照，CNY）。 */
export interface ApiKbUsageResponse {
  sourceId: number | null
  dateFrom: string | null
  dateTo: string | null
  currency: string
  tokenItems: ApiKbUsageTokenItem[]
  asr: ApiKbUsageAsr
  cleanTokensPredicted: number
  cleanTokensActual: number
  totalCost: number
}

File: backend/app/models/kb.py
"""知识库域的 SQLAlchemy ORM 模型（arch/12）。

知识源（kb_source）、素材（kb_media）、内容分段（kb_transcript_segment）、
知识点（kb_knowledge_point）、图片资产（kb_image_asset）与域设置单行
（kb_settings）。
PG 是唯一存储，检索列（embedding halfvec 向量 +
pg_trgm 词面生成列）同库内嵌，无投影层。
"""

from datetime import datetime
from typing import Any

from pgvector.sqlalchemy import HALFVEC
from sqlalchemy import (
    JSON,
    BigInteger,
    Boolean,
    DateTime,
    ForeignKey,
    Index,
    Integer,
    String,
    Text,
    text,
)
from sqlalchemy.dialects.postgresql import JSONB
from sqlalchemy.orm import Mapped, mapped_column

from app.constants.kb import KB_EMBEDDING_DIMS, KbDescribeStatus, KbProcessStatus
from app.core.clock import utc_now
from app.core.database import Base

_JSONB = JSONB().with_variant(JSON(), "sqlite")
# sqlite（单测）侧退化为 JSON 文本存 list，PG 侧走 pgvector 真实类型
_HALFVEC = HALFVEC(KB_EMBEDDING_DIMS).with_variant(JSON(), "sqlite")


class KbSource(Base):
    """知识源（知识库实例）：课程或电子书集合。"""

    __tablename__ = "kb_source"
    __table_args__ = (Index("idx_kb_source_active", "deleted_at"),)

    id: Mapped[int] = mapped_column(primary_key=True, autoincrement=True)
    source_type: Mapped[str] = mapped_column(String(16), nullable=False)
    name: Mapped[str] = mapped_column(String(200), nullable=False)
    author: Mapped[str | None] = mapped_column(String(100), nullable=True)
    description: Mapped[str | None] = mapped_column(Text, nullable=True)
    enabled: Mapped[bool] = mapped_column(Boolean, nullable=False, default=True)
    chapter_tree: Mapped[dict[str, Any]] = mapped_column(
        _JSONB, nullable=False, default=lambda: {"draft": None, "published": None}
    )
    storage_bytes: Mapped[int] = mapped_column(BigInteger, nullable=False, default=0)
    pending_cleanup_bytes: Mapped[int] = mapped_column(BigInteger, nullable=False, default=0)
    deleted_at: Mapped[datetime | None] = mapped_column(DateTime(timezone=True), nullable=True)
    created_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True), default=utc_now, nullable=False
    )
    updated_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True), default=utc_now, onupdate=utc_now, nullable=False
    )


class KbMedia(Base):
    """素材：课程的一集视频/音频，或一部书 PDF。"""

    __tablename__ = "kb_media"
    __table_args__ = (
        # 集号唯一仅约束课程（书的 episode_no 为 NULL；PG 侧为部分唯一索引，见迁移）
        Index("uq_kb_media_source_episode", "source_id", "episode_no", unique=True),
        # 哈希去重仅约束存活行（软删行不占哈希位，24h 恢复窗内可重传）
        Index(
            "uq_kb_media_source_hash",
            "source_id",
            "file_hash",
            unique=True,
            postgresql_where=text("deleted_at IS NULL"),
            sqlite_where=text("deleted_at IS NULL"),
        ),
        Index("idx_kb_media_source_status", "source_id", "process_status"),
        Index("idx_kb_media_deleted", "deleted_at"),
    )

    id: Mapped[int] = mapped_column(primary_key=True, autoincrement=True)
    source_id: Mapped[int] = mapped_column(
        ForeignKey("kb_source.id", ondelete="CASCADE"), nullable=False
    )
    media_kind: Mapped[str] = mapped_column(String(8), nullable=False)
    episode_no: Mapped[int | None] = mapped_column(Integer, nullable=True)
    title: Mapped[str] = mapped_column(String(300), nullable=False)
    file_name: Mapped[str] = mapped_column(String(500), nullable=False)
    relative_path: Mapped[str | None] = mapped_column(String(500), nullable=True)
    cos_key: Mapped[str] = mapped_column(String(500), nullable=False)
    file_size: Mapped[int] = mapped_column(BigInteger, nullable=False, default=0)
    file_hash: Mapped[str] = mapped_column(String(64), nullable=False)
    duration_seconds: Mapped[int | None] = mapped_column(Integer, nullable=True)
    page_count: Mapped[int | None] = mapped_column(Integer, nullable=True)
    process_status: Mapped[str] = mapped_column(String(16), nullable=False, default=KbProcessStatus.UPLOADED)
    process_error: Mapped[str | None] = mapped_column(String(1000), nullable=True)
    process_meta: Mapped[dict[str, Any]] = mapped_column(_JSONB, nullable=False, default=dict)
    extracted_at: Mapped[datetime | None] = mapped_column(DateTime(timezone=True), nullable=True)
    edited_at: Mapped[datetime | None] = mapped_column(DateTime(timezone=True), nullable=True)
    vision_at: Mapped[datetime | None] = mapped_column(DateTime(timezone=True), nullable=True)
    deleted_at: Mapped[datetime | None] = mapped_column(DateTime(timezone=True), nullable=True)
    created_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True), default=utc_now, nullable=False
    )
    updated_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True), default=utc_now, onupdate=utc_now, nullable=False
    )


class KbTranscriptSegment(Base):
    """内容分段：课程句级时间码区间 / 书页区间的一段文本。"""

    __tablename__ = "kb_transcript_segment"
    __table_args__ = (
        Index("uq_kb_segment_media_seq", "media_id", "seq_no", unique=True),
        Index("idx_kb_segment_dirty", "embedding_dirty"),
        Index("idx_kb_segment_source_dirty", "source_id", "embedding_dirty"),
    )

    id: Mapped[int] = mapped_column(primary_key=True, autoincrement=True)
    media_id: Mapped[int] = mapped_column(
        ForeignKey("kb_media.id", ondelete="CASCADE"), nullable=False
    )
    source_id: Mapped[int] = mapped_column(
        ForeignKey("kb_source.id", ondelete="CASCADE"), nullable=False
    )
    seq_no: Mapped[int] = mapped_column(Integer, nullable=False)
    text: Mapped[str] = mapped_column(Text, nullable=False)
    start_ms: Mapped[int | None] = mapped_column(BigInteger, nullable=True)
    end_ms: Mapped[int | None] = mapped_column(BigInteger, nullable=True)
    page_start: Mapped[int | None] = mapped_column(Integer, nullable=True)
    page_end: Mapped[int | None] = mapped_column(Integer, nullable=True)
    embedding: Mapped[list[float] | None] = mapped_column(_HALFVEC, nullable=True)
    embedding_dirty: Mapped[bool] = mapped_column(Boolean, nullable=False, default=True)
    created_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True), default=utc_now, nullable=False
    )
    updated_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True), default=utc_now, onupdate=utc_now, nullable=False
    )


class KbKnowledgePoint(Base):
    """知识点卡片：检索与引用的最小知识单元。"""

    __tablename__ = "kb_knowledge_point"
    __table_args__ = (
        Index("idx_kb_point_source_status", "source_id", "status"),
        Index("idx_kb_point_dirty", "embedding_dirty"),
    )

    id: Mapped[int] = mapped_column(primary_key=True, autoincrement=True)
    source_id: Mapped[int] = mapped_column(
        ForeignKey("kb_source.id", ondelete="CASCADE"), nullable=False
    )
    media_id: Mapped[int] = mapped_column(
        ForeignKey("kb_media.id", ondelete="CASCADE"), nullable=False
    )
    point_type: Mapped[str] = mapped_column(String(16), nullable=False)
    title: Mapped[str] = mapped_column(String(300), nullable=False)
    body: Mapped[str] = mapped_column(Text, nullable=False)
    term_definition: Mapped[str | None] = mapped_column(Text, nullable=True)
    applicable_scene: Mapped[str | None] = mapped_column(Text, nullable=True)
    excerpt: Mapped[str] = mapped_column(Text, nullable=False)
    start_ms: Mapped[int | None] = mapped_column(BigInteger, nullable=True)
    end_ms: Mapped[int | None] = mapped_column(BigInteger, nullable=True)
    page_start: Mapped[int | None] = mapped_column(Integer, nullable=True)
    page_end: Mapped[int | None] = mapped_column(Integer, nullable=True)
    related_ids: Mapped[list[Any]] = mapped_column(_JSONB, nullable=False, default=list)
    chapter_path: Mapped[list[Any]] = mapped_column(_JSONB, nullable=False, default=list)
    status: Mapped[str] = mapped_column(String(16), nullable=False, default="draft")
    needs_review: Mapped[bool] = mapped_column(Boolean, nullable=False, default=False)
    review_note: Mapped[str | None] = mapped_column(String(1000), nullable=True)
    reviewed_by: Mapped[int | None] = mapped_column(
        ForeignKey("user.id", ondelete="SET NULL"), nullable=True
    )
    reviewed_at: Mapped[datetime | None] = mapped_column(DateTime(timezone=True), nullable=True)
    embedding: Mapped[list[float] | None] = mapped_column(_HALFVEC, nullable=True)
    embedding_dirty: Mapped[bool] = mapped_column(Boolean, nullable=False, default=True)
    created_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True), default=utc_now, nullable=False
    )
    updated_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True), default=utc_now, onupdate=utc_now, nullable=False
    )


class KbImageAsset(Base):
    """图片资产：书嵌图与课程视频关键帧统一落表（文字搜图的检索载体）。"""

    __tablename__ = "kb_image_asset"
    __table_args__ = (
        Index("idx_kb_image_source_dirty", "source_id", "embedding_dirty"),
    )

    id: Mapped[int] = mapped_column(primary_key=True, autoincrement=True)
    media_id: Mapped[int] = mapped_column(
        ForeignKey("kb_media.id", ondelete="CASCADE"), nullable=False
    )
    source_id: Mapped[int] = mapped_column(
        ForeignKey("kb_source.id", ondelete="CASCADE"), nullable=False
    )
    page_no: Mapped[int | None] = mapped_column(Integer, nullable=True)
    start_ms: Mapped[int | None] = mapped_column(BigInteger, nullable=True)
    end_ms: Mapped[int | None] = mapped_column(BigInteger, nullable=True)
    bbox: Mapped[dict[str, Any] | None] = mapped_column(_JSONB, nullable=True)
    cos_key: Mapped[str] = mapped_column(String(500), nullable=False)
    thumb_cos_key: Mapped[str | None] = mapped_column(String(500), nullable=True)
    text_in_image: Mapped[str | None] = mapped_column(Text, nullable=True)
    caption: Mapped[str | None] = mapped_column(Text, nullable=True)
    vision_description: Mapped[str | None] = mapped_column(Text, nullable=True)
    describe_status: Mapped[str] = mapped_column(
        String(16), nullable=False, default=KbDescribeStatus.PENDING
    )
    describe_attempts: Mapped[int] = mapped_column(Integer, nullable=False, default=0)
    embedding: Mapped[list[float] | None] = mapped_column(_HALFVEC, nullable=True)
    embedding_dirty: Mapped[bool] = mapped_column(Boolean, nullable=False, default=True)
    index_excluded: Mapped[bool] = mapped_column(Boolean, nullable=False, default=False)
    created_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True), default=utc_now, nullable=False
    )
    updated_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True), default=utc_now, onupdate=utc_now, nullable=False
    )


class KbSettings(Base):
    """知识库域设置（单行）：域参数 + 模型角色槽位（llm_config 引用）。"""

    __tablename__ = "kb_settings"

    id: Mapped[int] = mapped_column(primary_key=True, autoincrement=True)
    hotwords: Mapped[list[Any]] = mapped_column(_JSONB, nullable=False, default=list)
    segment_max_seconds: Mapped[int] = mapped_column(Integer, nullable=False, default=30)
    asr_concurrency: Mapped[int] = mapped_column(Integer, nullable=False, default=2)
    top_k: Mapped[int] = mapped_column(Integer, nullable=False, default=8)
    auto_approve_points: Mapped[bool] = mapped_column(
        Boolean, nullable=False, default=True, server_default=text("true")
    )
    unit_prices: Mapped[dict[str, Any]] = mapped_column(_JSONB, nullable=False, default=dict)
    embedding_config_id: Mapped[int | None] = mapped_column(
        ForeignKey("llm_config.id", ondelete="SET NULL"), nullable=True
    )
    clean_model_id: Mapped[int | None] = mapped_column(
        ForeignKey("llm_config.id", ondelete="SET NULL"), nullable=True
    )
    extract_model_id: Mapped[int | None] = mapped_column(
        ForeignKey("llm_config.id", ondelete="SET NULL"), nullable=True
    )
    vision_model_id: Mapped[int | None] = mapped_column(
        ForeignKey("llm_config.id", ondelete="SET NULL"), nullable=True
    )
    authorized_user_ids: Mapped[list[Any]] = mapped_column(_JSONB, nullable=False, default=list)
    updated_by: Mapped[int | None] = mapped_column(
        ForeignKey("user.id", ondelete="SET NULL"), nullable=True
    )
    created_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True), default=utc_now, nullable=False
    )
    updated_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True), default=utc_now, onupdate=utc_now, nullable=False
    )
